config/v1/types_securityprofile.go

```diff
@@ -0,0 +1,304 @@
+	// TLS versions 1.1, 1.2 and 1.3 (yaml):
+	//
+	//   securityProtocol:
+	//     minimumVersion: TLSv1.1
+	//     maximumVersion: TLSv1.3
+	//
+	SecurityProtocol SecurityProtocol `json:"securityProtocol"`
+	// dhParamSize sets the maximum size of the Diffie-Hellman parameters used for generating
+	// the ephemeral/temporary Diffie-Hellman key in case of DHE key exchange. The final size
+	// will try to match the size of the server's RSA (or DSA) key (e.g, a 2048 bits temporary
+	// DH key for a 2048 bits RSA key), but will not exceed this maximum value.
+	//
+	// Available DH Parameter sizes are:
+	//
+	//   "2048": A Diffie-Hellman parameter of 2048 bits.
+	//   "1024": A Diffie-Hellman parameter of 1024 bits.
+	//
+	// For example, to use a Diffie-Hellman parameter of 2048 bits (yaml):
+	//
+	//   dhParamSize: 2048
+	//
+	DHParamSize DHParamSize `json:"dhParamSize"`
+}
+
+// SecurityProtocol defines one or more security protocols used to secure network connections.
+type SecurityProtocol struct {
+	// minimumVersion enforces use of the specified SecurityProtocolVersion or newer
+	// on SSL connections. minimumVersion must be lower than or equal to maximumVersion.
+	//
+	// If unset and maximumVersion is set, minimumVersion will be set
+	// to maximumVersion. If minimumVersion and maximumVersion are unset,
+	// the minimum version is determined by the security profile type.
+	//
+	//   SecurityProfileType Modern:       SecurityProtocolTLS12Version
+	//   SecurityProfileType Intermediate: SecurityProtocolTLS10Version
+	//   SecurityProfileType Old:          SecurityProtocolTLS10Version
+	//
+	// Supported minimum versions are:
+	//
+	//   "TLSv1.3": Version 1.3 of the TLS security protocol used for securing network connections.
+	//   "TLSv1.2": Version 1.2 of the TLS security protocol used for securing network connections.
+	//   "TLSv1.1": Version 1.1 of the TLS security protocol used for securing network connections.
+	//   "TLSv1.0": Version 1.0 of the TLS security protocol used for securing network connections.
+	//
+	MinimumVersion SecurityProtocolVersion `json:"minimumVersion"`
+	// maximumVersion enforces use of the specified SecurityProtocolVersion or older
+	// on SSL connections. maximumVersion must be higher than or equal to minimumVersion.
+	//
+	// If unset and minimumVersion is set, maximumVersion will be set
+	// to minimumVersion. If minimumVersion and maximumVersion are unset,
+	// the maximum version is determined by the security profile type.
+	//
+	//   SecurityProfileType Modern:       SecurityProtocolTLS12Version
+	//   SecurityProfileType Intermediate: SecurityProtocolTLS12Version
+	//   SecurityProfileType Old:          SecurityProtocolTLS12Version
+	//
+	// Supported maximum versions are the same as minimum versions.
+	//
+	MaximumVersion SecurityProtocolVersion `json:"maximumVersion"`
+}
+
+// SecurityProtocolVersion is a way to specify the TLS security protocol version used to
+// secure network connections.
```

Comment:
What codebase defines what the valid constants are?  OpenSSL?

Reply:
I created the constants from the haproxy config ssl min/max versions:
https://cbonte.github.io/haproxy-dconv/1.8/configuration.html#ssl-min-ver%20%28Bind%20options%29

Reply:
Add that to your godoc (specifically identify why these constants were chosen and who defines them).

I'm a bit concerned that this may even be too specific to HAProxy.  Can you verify that nginx / other openssl projects use similar constants?

Reply:
I have verified most other implementations use these constants by using https://ssl-config.mozilla.org/. I have updated the godoc accordingly.